feat: introduce new deprecated types for rules

DMartens opened the pull request:
<!--
    Thank you for contributing!

    ESLint adheres to the [OpenJS Foundation Code of Conduct](https://eslint.org/conduct).
-->

#### Prerequisites checklist

- [x] I have read the [contributing guidelines](https://github.com/eslint/eslint/blob/HEAD/CONTRIBUTING.md).

#### What is the purpose of this pull request? (put an "X" next to an item)

<!--
    The following template is intentionally not a markdown checkbox list for the reasons
    explained in https://github.com/eslint/eslint/pull/12848#issuecomment-580302888
-->

[x] Documentation update
[ ] Bug fix ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/bug-report.md))
[ ] New rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-proposal.md))
[ ] Changes an existing rule ([template](https://raw.githubusercontent.com/eslint/eslint/HEAD/templates/rule-change-proposal.md))
[ ] Add autofix to a rule
[ ] Add a CLI option
[x] Add something to the core
[ ] Other, please explain:

<!--
    If the item you've checked above has a template, please paste the template questions below and answer them. (If this pull request is addressing an issue, you can just paste a link to the issue here instead.)
-->

<!--
    Please ensure your pull request is ready:

    - Read the pull request guide (https://eslint.org/docs/latest/contribute/pull-requests)
    - Include tests for this change
    - Update documentation for this change (if appropriate)
-->

<!--
    The following is required for all pull requests:
-->

#### What changes did you make? (Give an overview)
This implements the [additional deprecated metadata RFC](https://github.com/eslint/rfcs/blob/02ba56c5ec0f84b8861fc70642b6b26181c73c45/designs/2024-deprecated-rule-metadata/README.md) and closes #18061.

The checklist from the RFC:
Internal Changes:
- [x] Mention the new schema in the custom rule documentation
- [x] Add any additional information to these properties in core rules as desired (such as in eslint/eslint#18053)
- [x] Update ESLint's website generator to take into account the additional information for rule doc deprecation notices
- [x] Update LintResult.usedDeprecatedRules by normalizing the old and new format for the existing replacedBy property and adding a new property with the name info for rules using the new deprecated format
- [x] Update the types in eslint

External Changes:
- [ ] Update the types in @typescript-eslint/eslint
- [ ] Update eslint-doc-generator to handle the new information: bmish/eslint-doc-generator#512
- [ ] Update the metadata for the most common plugins
- [ ] Consider implementing an eslint-plugin-eslint-plugin rule to encourage more complete deprecation information to be stored in these properties

I also investigated the impact for third-party plugins.
Generally there are only a few deprecated rules and most are stylistic rules.
An overview for plugins with some deprecated rules:
- [@typescript-eslint](https://github.com/typescript-eslint/typescript-eslint): [Several](https://typescript-eslint.io/rules/?=deprecated)
- [react](https://github.com/jsx-eslint/eslint-plugin-react): `jsx-*`
- [node](https://github.com/eslint-community/eslint-plugin-n): `shebang`
- [import](https://github.com/import-js/eslint-plugin-import): `imports-first`
- [unicorn](https://github.com/sindresorhus/eslint-plugin-unicorn): [Several](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/main/docs/deprecated-rules.md)
- [vue](https://github.com/vuejs/eslint-plugin-vue) : `component-tags-order`, `no-invalid-model-keys`, `no-ref-object-destructure`, `no-setup-props-destructure`, `script-setup-uses-vars`, `v-on-function-call`
- [svelte](https://github.com/sveltejs/eslint-plugin-svelte): `svelte/@typescript-eslint/no-unnecessary-condition`
- [functional](https://github.com/eslint-functional/eslint-plugin-functional):  `functional-parameters`

Plugins without deprecated rules:
- [jsdoc](https://github.com/gajus/eslint-plugin-jsdoc)
- [jest](https://github.com/jest-community/eslint-plugin-jest)
- [astro](https://github.com/ota-meshi/eslint-plugin-astro)
- [eslint-comments](https://github.com/eslint-community/eslint-plugin-eslint-comments)
- [eslint-plugin](https://github.com/eslint-community/eslint-plugin-eslint-plugin)
- [promise](https://github.com/eslint-community/eslint-plugin-promise) (deprecated rules deleted)
- [mocha](https://github.com/lo1tuma/eslint-plugin-mocha)
- [regexp](https://github.com/ota-meshi/eslint-plugin-regexp) (deprecated rules deleted)
- [security](https://github.com/eslint-community/eslint-plugin-security)

#### Is there anything you'd like reviewers to focus on?

Some notes for the changes:
- I did not know where the put the documentation for `DeprecateInfo` as it is part of the rule but its documentation has no type anywhere and the type documentation is required for `usedDeprecatedRules` in NodeJS API Documentation. I chose to use a separate page with a short intro about the deprecated format and detailing the new format. This document could also detail future tools using the deprecated information.
-  `meta.deprecated` property for the rule `indent-legacy` is not updated as it is used for a test case in `legacy-eslint` and `CLIEngine` which I assume to be frozen (but I still updated it for other frozen rules)
- the replacements only mentions the latest replacement in `eslint-stylistic` (e.g. rules replaced by  `padding-line-between-statements`)

<!-- markdownlint-disable-file MD004 -->

mdjermanovic commented on lib/rules/array-bracket-newline.js:
```diff
@@ -15,8 +15,26 @@ const astUtils = require("./utils/ast-utils");
-        deprecated: true,
-        replacedBy: [],
+        deprecated: {
+            message: "Stylistic rules are being moved out of ESLint core.",
```
```suggestion
            message: "Formatting rules are being moved out of ESLint core.",
```

We'll still have stylistic but not formatting rules in ESLint core.

(the same suggestion for all other rules with this message)

mdjermanovic commented on lib/rules/array-bracket-newline.js:
```diff
@@ -15,8 +15,26 @@ const astUtils = require("./utils/ast-utils");
+            availableUntil: "10.0.0",
+            replacedBy: [
+                {
+                    message: "Maintains now deprecated stylistic core rules",
```
```suggestion
                    message: "ESLint Stylistic maintains now deprecated stylistic core rules",
```

Since there's no other text (just URLs) for this replacement entry, I think we should mention the project's name in this message.

mdjermanovic commented:
> * Update ESLint's website generator to take into account the additional information for rule doc deprecation notices

It doesn't seem that this is implemented?

mdjermanovic commented:
> This implements the [additional deprecated metadata RFC](https://github.com/eslint/rfcs/blob/02ba56c5ec0f84b8861fc70642b6b26181c73c45/designs/2024-deprecated-rule-metadata/README.md) and closes #18053.

I think this closes https://github.com/eslint/eslint/issues/18061 but not #18053.

DMartens commented:
I updated the OP with the correct issue and corrected the messages in the rules.
Regarding the ESLint website update:
As noted in the notes I did not need to update the website code to keep working (it looks the same as without the changes).
Which additional details should be shown on the rule pages?

nzakas commented:
I think what we're looking for is something to autogenerate the comment at the top of this page:
https://eslint.org/docs/latest/rules/comma-style

> This rule was deprecated in ESLint v8.53.0. Please use the [corresponding rule](https://eslint.style/rules/js/comma-style) in [@stylistic/eslint-plugin-js](https://eslint.style/packages/js).

Right now, these are hardcoded into the docs.

mdjermanovic commented:
> As noted in the notes I did not need to update the website code to keep working (it looks the same as without the changes).
> Which additional details should be shown on the rule pages?

Sorry for the delay, I somehow missed a notification for this comment.

My understanding of the "Update ESLint's website generator to take into account the additional information" task mentioned in the original issue and the RFC was that we'll use the new metadata to autogenerate deprecation text on rule pages instead of hardcoding it in each rule's docs source .md file, just as @nzakas said in the comment above.

DMartens commented:
Sorry for the delay. I added the deprecated rule notice and removed the manually added ones.
I chose an important alert to make it consistent with the removed rules (e.g. `require-jsdoc`).

mdjermanovic commented:
@DMartens can you please resolve merge conflicts?

mdjermanovic commented:
Verify Files check is failing:

<details>

```
Validating rules
Missing deprecated information in yield-star-spacing rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in wrap-regex rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in wrap-iife rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in template-tag-spacing rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in template-curly-spacing rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in switch-colon-spacing rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in spaced-comment rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in space-unary-ops rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in space-infix-ops rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in space-in-parens rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in space-before-function-paren rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in space-before-blocks rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in semi rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in semi-style rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in semi-spacing rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in rest-spread-spacing rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in quotes rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in quote-props rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in prefer-reflect rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in padding-line-between-statements rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in padded-blocks rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in operator-linebreak rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in one-var-declaration-per-line rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in object-property-newline rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in object-curly-spacing rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in object-curly-newline rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in nonblock-statement-body-position rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in no-whitespace-before-property rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in no-trailing-spaces rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in no-tabs rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in no-sync rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in no-spaced-func rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in no-return-await rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in no-restricted-modules rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in no-process-exit rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in no-process-env rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in no-path-concat rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in no-new-symbol rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in no-new-require rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in no-negated-in-lhs rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in no-native-reassign rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in no-multiple-empty-lines rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in no-multi-spaces rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in no-mixed-spaces-and-tabs rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in no-mixed-requires rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in no-mixed-operators rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in no-floating-decimal rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in no-extra-semi rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in no-extra-parens rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in no-confusing-arrow rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in no-catch-shadow rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in no-buffer-constructor rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in newline-per-chained-call rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in newline-before-return rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in newline-after-var rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in new-parens rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in multiline-ternary rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in multiline-comment-style rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in max-statements-per-line rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in max-len rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in lines-between-class-members rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in lines-around-directive rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in lines-around-comment rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in linebreak-style rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in line-comment-position rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in keyword-spacing rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in key-spacing rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in jsx-quotes rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in indent rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in indent-legacy rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in implicit-arrow-linebreak rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in id-blacklist rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in handle-callback-err rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in global-require rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in generator-star-spacing rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in function-paren-newline rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in function-call-argument-newline rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in func-call-spacing rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in eol-last rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in dot-location rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in computed-property-spacing rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in comma-style rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in comma-spacing rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in comma-dangle rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in callback-return rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in brace-style rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in block-spacing rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in arrow-spacing rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in arrow-parens rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in array-element-newline rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in array-bracket-spacing rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Missing deprecated information in array-bracket-newline rule code or README.md. Please write @deprecated tag in code and「This rule was deprecated in ESLint ...」 in README.md.
Error: Process completed with exit code 1.
```

</details>

 Looks like we should remove checking for deprecated info in rule docs source files.

https://github.com/eslint/eslint/blob/17f2aaec16d5afbb0d219bce6ae01d7b15d74828/Makefile.js#L752-L757

mdjermanovic commented on docs/src/extend/custom-rules.md:
```diff
@@ -66,9 +66,10 @@ The source file for a rule exports an object with the following properties. Both
 
-* `deprecated`: (`boolean`) Indicates whether the rule has been deprecated.  You may omit the `deprecated` property if the rule has not been deprecated.
+* `deprecated`: (`boolean | DeprecatedInfo`) Indicates whether the rule has been deprecated.  You may omit the `deprecated` property if the rule has not been deprecated.  
+There is a dedicated page for the [DeprecatedInfo](./rule-deprecation.md#Type)
```
```suggestion
There is a dedicated page for the [DeprecatedInfo](./rule-deprecation)
```

mdjermanovic commented on docs/src/integrate/nodejs-api.md:
```diff
@@ -406,8 +406,9 @@ The `LintResult` value is the information of the linting result of each file. Th
-* `usedDeprecatedRules` (`{ ruleId: string; replacedBy: string[] }[]`)<br>
+* `usedDeprecatedRules` (`{ ruleId: string; replacedBy: string[]; info: DeprecatedInfo | undefined }[]`)<br>
   The information about the deprecated rules that were used to check this file.
+  The info property is defined to `rule.meta.deprecated` if the rule uses the [new deprecated property](../extend/rule-deprecation.md).
```
```suggestion
  The `info` property is set to `rule.meta.deprecated` if the rule uses the [new `deprecated` property](../extend/rule-deprecation).
```

fasttime commented on docs/src/extend/rule-deprecation.md:
```diff
@@ -0,0 +1,49 @@
+  [Semver](https://semver.org/) of the version deprecating the rule.
+* `availableUntil` (`string | null`)<br>
+  [Semver](https://semver.org/) of the version likely to remove the rule, e.g. the next major version.
+  The special value `null` means the rule is "frozen". This means the rule will no longer be changed but will be kept available.
```
[The RFC](https://github.com/eslint/rfcs/blob/main/designs/2024-deprecated-rule-metadata/README.md?plain=1#L62) does indeed state that a value of `null` for `deprecated.availableUntil` will indicate 
a frozen rule, but the new implementation doesn't account for that use case. If we want to account for frozen rules in the implementation then I think we should:

1. Add a `meta.deprecated` property to frozen rules, with `meta.deprecated.availableUntil` set to `null`.
2. Update the way we check for deprecated rules (i.e. rules that are deprecated but not frozen). Instead of testing just the truthiness of `meta.deprecated`, we should also test that `meta.deprecated.availableUntil` isn't `null`. There are a few places where this kind of check is performed, both in production code and in the build, and each check would have to be updated.

I'm not sure if this is what we want, so I wouldn't be opposed to leaving out the mention of the special `null` value from the rule deprecation docs and ignoring the case of frozen rules for now. Thoughts? @eslint/eslint-tsc

mdjermanovic replied:
My understanding of "frozen" (`availableUntil:null`) in the RFC was that the rule is deprecated but won't be removed. We're using the term "frozen" for rules that are not deprecated but are not accepting new feature requests, so I don't think we should add `meta.deprecated` to those rules since they are not deprecated.

I think it would be best to remove the word "frozen" from this description to avoid confusion:

```suggestion
  The special value `null` means the rule will no longer be changed but will be kept available.
```

fasttime commented on lib/shared/types.js:
```diff
@@ -177,6 +194,7 @@ module.exports = {};
  * @typedef {Object} DeprecatedRuleInfo
  * @property {string} ruleId The rule ID.
  * @property {string[]} replacedBy The rule IDs that replace this deprecated rule.
+ * @property {DeprecatedInfo} [info] The raw deprecated info provided by rule. Unset if deprecated is a boolean.
```
```suggestion
 * @property {DeprecatedInfo} [info] The raw deprecated info provided by the rule. Unset if `deprecated` is a boolean.
```

fasttime commented on lib/rules/prefer-reflect.js:
```diff
@@ -20,9 +20,13 @@ module.exports = {
-
-        replacedBy: [],
+        deprecated: {
+            message: "Original intentation was misguided",
```
```suggestion
            message: "Original intention was misguided",
```

fasttime commented on lib/rules/global-require.js:
```diff
@@ -52,9 +52,25 @@ function isShadowed(scope, node) {
-
-        replacedBy: [],
+        deprecated: {
+            message: "Node rule were moved out of ESLint core.",
```
Maybe "Node.js" sounds better than just "Node" here. More importantly, the sentence should read "Node.js rule**s** were ..." or "Node.js rule **was** ...". We should correct this message in multiple places.

fasttime commented on lib/rules/global-require.js:
```diff
@@ -52,9 +52,25 @@ function isShadowed(scope, node) {
+            availableUntil: null,
+            replacedBy: [
+                {
+                    message: "ESLint Stylistic Maintains now deprecated node rules",
```
```suggestion
                    message: "ESLint Stylistic maintains now deprecated node rules",
```

In other places as well.

fasttime commented on lib/rules/array-bracket-newline.js:
```diff
@@ -15,8 +15,26 @@ const astUtils = require("./utils/ast-utils");
+            availableUntil: "10.0.0",
+            replacedBy: [
+                {
+                    message: "ESLint Stylistic Maintains now deprecated stylistic core rules",
```
```suggestion
                    message: "ESLint Stylistic maintains now deprecated stylistic core rules.",
```

We have this message in other places as well.

nzakas commented:
@DMartens please check the CI failure.

fasttime commented on lib/rules/no-mixed-requires.js:
```diff
@@ -13,9 +13,25 @@
+                    },
+                    rule: {
+                        name: "no-mixed-requires",
+                        url: "https://github.com/eslint-community/eslint-plugin-n/tree/master/docs/rules/no-mixed-requires.md"
```
```suggestion
                        url: "https://github.com/eslint-community/eslint-plugin-n/tree/master/docs/rules/no-mixed-requires"
```

fasttime commented on lib/rules/no-new-require.js:
```diff
@@ -13,9 +13,25 @@
+                    },
+                    rule: {
+                        name: "no-new-require",
+                        url: "https://github.com/eslint-community/eslint-plugin-n/tree/master/docs/rules/no-new-require.md"
```
```suggestion
                        url: "https://github.com/eslint-community/eslint-plugin-n/tree/master/docs/rules/no-new-require"
```

fasttime commented on lib/rules/no-path-concat.js:
```diff
@@ -12,9 +12,25 @@
+                    },
+                    rule: {
+                        name: "no-path-concat",
+                        url: "https://github.com/eslint-community/eslint-plugin-n/tree/master/docs/rules/no-path-concat.md"
```
```suggestion
                        url: "https://github.com/eslint-community/eslint-plugin-n/tree/master/docs/rules/no-path-concat"
```

fasttime commented on lib/rules/no-process-env.js:
```diff
@@ -12,9 +12,25 @@
+                    },
+                    rule: {
+                        name: "no-process-env",
+                        url: "https://github.com/eslint-community/eslint-plugin-n/tree/master/docs/rules/no-process-env.md"
```
```suggestion
                        url: "https://github.com/eslint-community/eslint-plugin-n/tree/master/docs/rules/no-process-env"
```

fasttime commented on lib/rules/no-process-exit.js:
```diff
@@ -12,9 +12,25 @@
+                    },
+                    rule: {
+                        name: "no-process-exit",
+                        url: "https://github.com/eslint-community/eslint-plugin-n/tree/master/docs/rules/no-process-exit.md"
```
```suggestion
                        url: "https://github.com/eslint-community/eslint-plugin-n/tree/master/docs/rules/no-process-exit"
```

fasttime commented on lib/rules/no-restricted-modules.js:
```diff
@@ -48,9 +48,25 @@ const arrayOfStringsOrObjects = {
+                    },
+                    rule: {
+                        name: "no-restricted-modules",
+                        url: "https://github.com/eslint-community/eslint-plugin-n/tree/master/docs/rules/no-restricted-modules.md"
```
```suggestion
                        url: "https://github.com/eslint-community/eslint-plugin-n/tree/master/docs/rules/no-restricted-modules"
```

fasttime commented on lib/rules/no-sync.js:
```diff
@@ -13,9 +13,25 @@
+                    },
+                    rule: {
+                        name: "no-sync",
+                        url: "https://github.com/eslint-community/eslint-plugin-n/tree/master/docs/rules/no-sync.md"
```
```suggestion
                        url: "https://github.com/eslint-community/eslint-plugin-n/tree/master/docs/rules/no-sync"
```

fasttime commented on lib/rules/no-buffer-constructor.js:
```diff
@@ -12,9 +12,25 @@
+                    },
+                    rule: {
+                        name: "no-buffer-constructor",
+                        url: "https://github.com/eslint-community/eslint-plugin-n/tree/master/docs/rules/no-buffer-constructor.md"
```
```suggestion
                        url: "https://github.com/eslint-community/eslint-plugin-n/tree/master/docs/rules/no-buffer-constructor"
```

fasttime commented on lib/rules/global-require.js:
```diff
@@ -52,9 +52,25 @@ function isShadowed(scope, node) {
+                    },
+                    rule: {
+                        name: "global-require",
+                        url: "https://github.com/eslint-community/eslint-plugin-n/tree/master/docs/rules/global-require"
```
It looks like the `.md` extension is required for the like to GitHub files (`https://github.com/`) to work. I'm sorry if last comment was misleading:

```suggestion
                        url: "https://github.com/eslint-community/eslint-plugin-n/tree/master/docs/rules/global-require.md"
```

fasttime commented:
[`callback-return.js`](https://github.com/eslint/eslint/blob/main/lib/rules/callback-return.js) and [`indent-legacy.js`](https://github.com/eslint/eslint/blob/main/lib/rules/indent-legacy.js) should be also switched to the new `deprecated` format. And we should remove the deprecation note from [`no-new-object.md`](https://github.com/eslint/eslint/blob/main/docs/src/rules/no-new-object.md).

nzakas commented on lib/rules/no-buffer-constructor.js:
```diff
@@ -12,9 +12,25 @@
+            availableUntil: null,
+            replacedBy: [
+                {
+                    message: "eslint-plugin-n now maintains deprecated Node.js-related rules",
```
```suggestion
                    message: "eslint-plugin-n now maintains deprecated Node.js-related rules.",
```

nzakas commented on lib/rules/array-bracket-newline.js:
```diff
@@ -15,8 +15,26 @@ const astUtils = require("./utils/ast-utils");
+            availableUntil: "10.0.0",
+            replacedBy: [
+                {
+                    message: "ESLint Stylistic now maintains deprecated stylistic core rules",
```
There's an inconsistency between the top-level `message`, which ends with a period, and this one, which does not. I think we should be consistent and have messages end with periods throughout.

mdjermanovic commented:
Links to replacement rules are now missing on the rules reference page. 

For example, on https://eslint.org/docs/latest/rules/:

![image](https://github.com/user-attachments/assets/9e957c0b-9ea4-4bfb-ae54-5efeb5098dbd)


On https://deploy-preview-19238--docs-eslint.netlify.app/rules/:

![image](https://github.com/user-attachments/assets/a34e407a-93d7-4d11-8b07-a0b4624f8a54)

fasttime commented:
> I did not update `indent-legacy` and `callback-return` as these are used in `LegacyESLint` s [tests](https://github.com/eslint/eslint/blob/208e0b199f5d5f9dd173e58e3f5db19c1f0c38ed/tests/lib/eslint/legacy-eslint.js#L1840C1-L1860C12). I assumed it is frozen and as such did not update its `usedDeprecatedRules` which would be required for changing to the new `deprecated` format.

I think it would be good to update `LegacyESLint` so that the format of `LintResult.usedDeprecatedRules` remains the same regardless of any changes we've made to the rules. Thoughts?

fasttime commented on Makefile.js:
```diff
@@ -199,7 +199,7 @@ function generateRuleIndexPage() {
                 ruleTypesData.deprecated.push({
                     name: basename,
-                    replacedBy: rule.meta.replacedBy || [],
+                    replacedBy: rule.meta.deprecated.replacedBy?.map(replacement => replacement.rule.name) ?? [],
```
The build is now failing because the autogenerated rules.json file (`docs/src/_data/rules.json`) contains the names of replacement rules that are in some other plugin, so the link checker fails because it finds links to pages that don't exist like `/docs/head/rules/function-call-spacing`.

If we only want to show replacement rules that are in the core as we are currenly doing then we should filter out rules that aren't in the core by checking that `replacement.plugin` is undefined.

nzakas replied:
I think it would be good to distinguish between core replacement rules and external. It's helpful to have links to the replacement rules in the docs, and if the deprecation information has a URL for that replacement rule, we should use that instead.

fasttime commented:
It's also fine to leave `LegacyESLint` unchanged as @nzakas suggests. This means that consumers of `LegacyESLint` will see `usedDeprecatedRules.replacedBy` as an empty array for rules that use the new deprecation format.

In this case we can update `callback-return.js` and `indent-legacy.js` to use the new `deprecated` format and update the tests for `LegacyESLint` to reflect the new expectation, i.e. check that `usedDeprecatedRules.replacedBy` is empty.

fasttime commented on Makefile.js:
```diff
@@ -199,7 +199,7 @@ function generateRuleIndexPage() {
                 ruleTypesData.deprecated.push({
                     name: basename,
-                    replacedBy: rule.meta.replacedBy || [],
+                    replacedBy: rule.meta.deprecated.replacedBy?.map(replacement => (replacement.rule.name !== "function-call-spacing" ? replacement.rule.name : "func-call-spacing")) ?? [],
```
We can't do this 😄 We could set the `rule.meta.deprecated.replacedBy` array without changes here, but we'll need to update the `rule.macro.html` template (docs/src/_includes/components/rule.macro.html) in order to render the links with the correct label and URL:

https://github.com/eslint/eslint/blob/613c06a2c341758739473409a2331074884ec7f8/docs/src/_includes/components/rule.macro.html#L12

https://github.com/eslint/eslint/blob/613c06a2c341758739473409a2331074884ec7f8/docs/src/_includes/components/rule.macro.html#L21

The above code uses the `ruleList` macro (docs/src/_includes/components/rule-categories.macro.html) assuming that `params.replacedBy` is an array of strings (i.e. names of core rules), but we will probably need a new macro in order to render external links.

DMartens replied:
The idea was the link consistently points to an internal page and then the user can navigate to the replacement via the deprecation notice.

I added a new macro for "external specifiers" but now the `ruleList` macro is unused. Should I remove it?

fasttime replied:
Since the new macro is being used for both external and built-in rules, I think we can remove the current `ruleList` implementation and rename `externalRuleList` in `ruleList`.  So in the end we'd be just changing the macro signature.

I'm not sure if it's sufficient to show just the name of the replacement rule in the notice as that looks a bit silly when it's the same name as the old rule.

![image](https://github.com/user-attachments/assets/04dd28ea-b0cc-43bf-acce-8a3096364044)

Maybe it would be better to include also the name of the plugin when the replacement rule is in a plugin, e.g.

> Replaced by [`no-confusing-arrow`](https://eslint.style/rules/js/no-confusing-arrow) in [@stylistic/eslint-plugin-js](https://eslint.style/packages/js)

Alternatively, we could include the standard prefix for the plugin in the text of the link:

> Replaced by [`@stylistic/js/no-confusing-arrow`](https://eslint.style/rules/js/no-confusing-arrow)

Thoughts?

DMartens replied:
I added the first variant as this makes it more obvious that the rule is provided in an external plugin.

fasttime commented:
Sorry for the delay @DMartens, I was busy the last days and forgot to review the changes. Could you please remove the deprecation notice from `no-new-object.md` docs as it's no longer needed?

https://github.com/eslint/eslint/blob/22a558228ff98f478fa308c9ecde361acc4caf20/docs/src/rules/no-new-object.md?plain=1#L9

nzakas commented:
It should have been removed rather than commented out.

fasttime commented:
> There is now a merge conflict because of a commented out [context.report type definition](https://github.com/eslint/eslint/blob/b476a930bb3a6d644c482747d985f5da0d89e1e9/lib/types/index.d.ts#L804). Is the comment intended?

Yeah, there have been a few changes around that line lately. Just go on and remove the comment.

fasttime commented on docs/src/_includes/components/rule-list.macro.html:
```diff
@@ -1,3 +1,10 @@
 {%- macro ruleList(params) -%}
     {% for rule in params.rules %}<a href="{{ ['/rules/', rule] | join | url }}" class="rule-list-item"><code>{{ rule }}</code></a>{% endfor %}
 {%- endmacro -%}
+
+{%- macro externalRuleList(params) -%}
+    {% for specifier in params.specifiers %}
+        <a href="{{ specifier.rule.url | url }}" class="rule-list-item"><code>{{ specifier.rule.name }}</code></a>
+        {% if specifier.plugin %}<span> in <a href="{{ specifier.plugin.url | url }}"><code>{{ specifier.plugin.name }}</code></a> {% endif %}
+    {% endfor %}
+{%- endmacro -%}
```
1. The name `externalRuleList` could be confusing since we are using this macro also for built-in rules. Maybe `replacementRuleList` would be better, or perhaps just `ruleList`, as the existing `ruleList` macro is no longer used.
2. We should also update the macro instructions in [`docs/src/library/rule-list.md`](https://github.com/eslint/eslint/blob/main/docs/src/library/rule-list.md).
3. If there are multiple replacement rules (in `params.specifiers`), the rule and plugin names will be displayed without a comma in between, which would break the formatting. This is not a problem currently since we are only showing the first replacement rule - because there are no rules with multiple replacements - but it would be good to add at least a comment in the macro to indicate that the current implementation is limited to showing one rule only.

DMartens replied:
The macro now handles multiple replacements with an or: ![example](https://github.com/user-attachments/assets/8fc25f1f-6e66-435c-9201-88f5db5a3dcb)

fasttime replied:
Thanks! Can you also update the macro title to "Replacement rule list" in the docs:

https://github.com/eslint/eslint/blob/2ae6aab993d6e584260e2de15234aecb077822d5/docs/src/library/rule-list.md?plain=1#L2

And indicate that the macro accepts a list of `ReplacedByInfo` objects and renders them as _or_-separated links:

https://github.com/eslint/eslint/blob/2ae6aab993d6e584260e2de15234aecb077822d5/docs/src/library/rule-list.md?plain=1#L5

fasttime commented on lib/types/index.d.ts:
```diff
@@ -765,6 +768,46 @@ export namespace Rule {
         hasSuggestions?: boolean | undefined;
     }
 
+    /** Provides additional metadata about a deprecation. */
+    interface DeprecatedInfo {
```
The types `DeprecatedInfo`, `ReplacedByInfo` and `ExternalSpecifier` are now included in `@eslint/core` and can be imported from there.

fasttime reviewed (commented):
I think this is good to go when the merge conflict is resolved.

nzakas commented on lib/eslint/eslint.js:
```diff
@@ -203,7 +203,13 @@ function getOrFindUsedDeprecatedRules(eslint, maybeFilePath) {
+
+                    retv.push({
+                        ruleId,
+                        replacedBy: usesNewFormat ? meta.deprecated.replacedBy?.map(replacement => `${replacement.plugin?.name !== void 0 ? `${replacement.plugin.name}/` : ""}${replacement.rule?.name ?? ""}`) ?? [] : meta.replacedBy || [],
```
I think we should use the name as it would likely appear in the config, so `@stylistic/js/semi` would be my choice.

DMartens replied:
Should the shorthand name also be used when displaying the replacement rules in the documentation?

fasttime reviewed (dismissed):
LGTM, thanks! Leaving open for other reviews.

mdjermanovic commented on Makefile.js:
```diff
@@ -744,18 +744,6 @@ target.checkRuleFiles = function() {
-            const deprecatedTagRegExp = /@deprecated in ESLint/u;
-            const deprecatedInfoRegExp = /This rule was .+deprecated.+in ESLint/u;
-
-            return deprecatedTagRegExp.test(ruleCode) && deprecatedInfoRegExp.test(docText);
```
Can we keep the check for `ruleCode`?

mdjermanovic commented on lib/rules/multiline-comment-style.js:
```diff
@@ -14,8 +14,26 @@ const astUtils = require("./utils/ast-utils");
+                        url: "https://eslint.style/packages/js"
+                    },
+                    rule: {
+                        name: "no-mixed-operators",
+                        url: "https://eslint.style/rules/js/no-mixed-operators"
```
```suggestion
                        name: "multiline-comment-style",
                        url: "https://eslint.style/rules/js/multiline-comment-style"
```

mdjermanovic commented on docs/src/_includes/layouts/doc.html:
```diff
@@ -90,6 +91,22 @@ <h1>{{ title }}</h1>
+                            {% elif rule_meta.deprecated.replacedBy[0].plugin %}
+                                Please use the <a href="{{ rule_meta.deprecated.replacedBy[0].rule.url }}">corresponding rule</a> in <a href="{{ rule_meta.deprecated.replacedBy[0].plugin.url }}">{{ rule_meta.deprecated.replacedBy[0].plugin.name }}</a>.
+                            {% else %}
+                                Please replace the rule with <a href="{{ rule.meta.deprecated.replacedBy[0].rule.url}}">{{ rule.meta.deprecated.replacedBy[0].rule.name }}</a>
```
This doesn't seem to work:

![image](https://github.com/user-attachments/assets/8102c40f-2826-4a8d-9ec8-9ae84e1d6497)

While at this, since the docs sites are versioned, I think the link should be relative (`href` could be just the replacement rule's name).

mdjermanovic commented on lib/rules/no-buffer-constructor.js:
```diff
@@ -12,9 +12,25 @@
+                        url: "https://github.com/eslint-community/eslint-plugin-n"
+                    },
+                    rule: {
+                        name: "no-buffer-constructor",
+                        url: "https://github.com/eslint-community/eslint-plugin-n/tree/master/docs/rules/no-buffer-constructor.md"
```
```suggestion
                        name: "no-deprecated-api",
                        url: "https://github.com/eslint-community/eslint-plugin-n/tree/master/docs/rules/no-deprecated-api.md"
```

Per https://github.com/mysticatea/eslint-plugin-node/pull/206#discussion_r377549508

mdjermanovic commented on lib/rules/no-restricted-modules.js:
```diff
@@ -48,9 +48,25 @@ const arrayOfStringsOrObjects = {
+                        url: "https://github.com/eslint-community/eslint-plugin-n"
+                    },
+                    rule: {
+                        name: "no-restricted-modules",
+                        url: "https://github.com/eslint-community/eslint-plugin-n/tree/master/docs/rules/no-restricted-modules.md"
```
```suggestion
                        name: "no-restricted-require",
                        url: "https://github.com/eslint-community/eslint-plugin-n/tree/master/docs/rules/no-restricted-require.md"
```

Per https://github.com/mysticatea/eslint-plugin-node/pull/206#discussion_r377552123

mdjermanovic commented on docs/src/_includes/layouts/doc.html:
```diff
@@ -90,6 +91,22 @@ <h1>{{ title }}</h1>
+                            {% elif rule_meta.deprecated.replacedBy[0].plugin %}
+                                Please use the <a href="{{ rule_meta.deprecated.replacedBy[0].rule.url }}">corresponding rule</a> in <a href="{{ rule_meta.deprecated.replacedBy[0].plugin.url }}">{{ rule_meta.deprecated.replacedBy[0].plugin.name }}</a>.
+                            {% else %}
+                                Please replace the rule with <a href="{{ rule_meta.deprecated.replacedBy[0].rule.name }}">{{ rule_meta.deprecated.replacedBy[0].rule.name }}</a>
```
```suggestion
                                Please replace the rule with <a href="{{ rule_meta.deprecated.replacedBy[0].rule.name }}">{{ rule_meta.deprecated.replacedBy[0].rule.name }}</a>.
```

mdjermanovic commented on docs/src/_includes/components/rule-list.macro.html:
```diff
@@ -1,3 +1,7 @@
-{%- endmacro -%}
+{%- macro replacementRuleList(params) -%}
+    {% for specifier in params.specifiers %}
+        <a href="{{ specifier.rule.url | url }}" class="rule-list-item"><code>{{ specifier.rule.name }}</code></a>
```
Can this also be a relative link (rule name) if the replacement is a core rule?

mdjermanovic commented on docs/src/_includes/components/rule.macro.html:
```diff
@@ -18,7 +18,7 @@
         </p>
         {%- if params.replacedBy -%}
-        <p class="rule__description">Replaced by {{ ruleList({ rules: params.replacedBy }) }}</p>
+        <p class="rule__description">Replaced by {{ replacementRuleList({ specifiers: params.replacedBy }) }}</p>
```
Replacement rules for removed ones are not rendered:

![image](https://github.com/user-attachments/assets/52c6247f-3e4d-486b-9a2b-42c10fef96e8)

https://deploy-preview-19238--docs-eslint.netlify.app/rules/#removed

I think it's because `replacedBy` is in the old format (array of strings). We could update the data in [`rule-type-list.json`](https://github.com/eslint/eslint/blob/main/conf/rule-type-list.json) to use the new format, e.g., `["generator-star-spacing"]` -> `[{ "rule": { "name": "generator-star-spacing" } }]`

DMartens commented:
Thanks for the helpful code review, all of these should be fixed now.
Also fixed the if condition if there are no replacement rules (e.g. `valid-jsdoc`),

mdjermanovic commented on docs/src/extend/rule-deprecation.md:
```diff
@@ -0,0 +1,52 @@
+* `name` (`string`)<br>
+  The package name for `plugin` and the rule id for `rule`.
+* `url` (`string`)<br>
+  An URL pointing to additional information, for example an announcement post or migration guide.
```
```suggestion
  An URL pointing to documentation for the plugin / rule..
```

mdjermanovic commented on lib/rules/no-spaced-func.js:
```diff
@@ -21,9 +21,26 @@ module.exports = {
+        deprecated: {
+            message: "Formatting rules are being moved out of ESLint core.",
+            url: "https://eslint.org/blog/2016/08/eslint-v3.3.0-released/#deprecated-rules",
+            deprecatedSince: "8.53.0",
```
```suggestion
            deprecatedSince: "3.3.0",
```

mdjermanovic commented on lib/rules/no-new-symbol.js:
```diff
@@ -21,11 +21,20 @@ module.exports = {
+            replacedBy: [
+                {
+                    rule: {
+                        name: "no-object-constructor",
+                        url: "https://eslint.org/docs/latest/rules/no-object-constructor"
```
```suggestion
                        name: "no-new-native-nonconstructor",
                        url: "https://eslint.org/docs/latest/rules/no-new-native-nonconstructor"
```

mdjermanovic reviewed (approved):
LGTM, thanks!
